analysis/src/CompletionBackEnd.ml

```diff
@@ -1448,14 +1488,50 @@ let rec getCompletionsForContextPath ~full ~opens ~rawOpens ~allFiles ~pos ~env
           |> getCompletionsForPath ~completionContext:Value ~exact:false
                ~package ~opens ~allFiles ~pos ~env ~scope
         in
-        completions
-        |> List.map (fun (completion : Completion.t) ->
-               {
-                 completion with
-                 name = completionName completion.name;
-                 env
-                 (* Restore original env for the completion after x->foo()... *);
-               })
+        let completions =
+          completions
+          |> List.map (fun (completion : Completion.t) ->
+                 {
+                   completion with
+                   name = completionName completion.name;
+                   env
+                   (* Restore original env for the completion after x->foo()... *);
+                 })
+        in
+        (* We add React element functions to the completion if we're in a JSX context *)
+        let forJsxCompletion =
+          match (inJsx, getTypePath typ) with
+          | true, Some (Path.Pident id) when Ident.name id = "int" -> Some "int"
+          | true, Some (Path.Pident id) when Ident.name id = "float" ->
+            Some "float"
+          | true, Some (Path.Pident id) when Ident.name id = "string" ->
+            Some "string"
+          | true, Some (Path.Pident id) when Ident.name id = "array" ->
+            (* Make sure the array contains React.element *)
+            let isReactElementArray =
+              match typ |> extractType ~env ~package with
+              | Some (Tarray (_env, payload)) -> (
+                match
+                  payload |> getTypePath |> Option.map pathIdentToString
+                with
+                | Some "React.element" -> true
+                | _ -> false)
+              | _ -> false
+            in
+            if isReactElementArray then Some "array" else None
```

Comment:
This only adds the array completions for arrays where we know for sure it contains `React.element`. The problem with that is that it requires compilation.

So, an alternative is to just add it to completions for all `array`, regardless of whether it's valid or not. That might produce a better experience.

Reply:
Perhaps the simpler version that kicks in for all arrays is less surprising -- one less difference to observe.

Reply:
✅